src/Microsoft.OData.Core/ODataWriterCore.cs

```diff
@@ -2459,22 +2459,81 @@ private ODataPath AppendEntitySetKeySegment(ODataPath odataPath, bool throwIfFai
-                    path = path.AddKeySegment(keys, currentEntityType, this.CurrentScope.NavigationSource);
+
+                    ODataResourceSerializationInfo serializationInfo = this.GetResourceSerializationInfo(resource);
+
+                    if (!throwIfFail)
+                    {
+                        if (resource.NonComputedProperties != null)
+                        {
+                            List<KeyValuePair<string, object>> keys = new List<KeyValuePair<string, object>>();
+
+                            if (serializationInfo != null)
+                            {
+                                foreach (ODataProperty property in resource.NonComputedProperties)
+                                {
+                                    if (property.SerializationInfo != null && property.SerializationInfo.PropertyKind == ODataPropertyKind.Key)
+                                    {
+                                        if(!CreateKeyValuePair(keys, property, property.Value))
+                                        {
+                                            return path;
+                                        }
+                                    }
+                                }
+                            }
+                            else
+                            {
+                                HashSet<string> keySet = new HashSet<string>();
+
+                                foreach (IEdmStructuralProperty property in currentEntityType.Key())
+                                {
+                                    keySet.Add(property.Name);
+                                }
+
+                                foreach (ODataProperty property in resource.NonComputedProperties)
+                                {
+                                    if (keySet.Contains(property.Name))
+                                    {
+                                        if (!CreateKeyValuePair(keys, property, property.Value))
+                                        {
+                                            return path;
+                                        }
+                                    }
+                                }
+                            }
+
+                            path = path.AddKeySegment(keys.ToArray(), currentEntityType, this.CurrentScope.NavigationSource);
+                        }                        
+                    }
+                    else 
+                    {
+                        KeyValuePair<string, object>[] keys = ODataResourceMetadataContext.GetKeyProperties(resource,
+                            this.GetResourceSerializationInfo(resource), currentEntityType);
+
+                        path = path.AddKeySegment(keys, currentEntityType, this.CurrentScope.NavigationSource);
+                    }                    
                 }
             }
             catch (ODataException)

```

Comment:
why still catch the exception if we just re-throw it? should just be able to remove the try/catch